pilot/copytool/rucio.py

```diff
@@ -76,35 +81,42 @@ def copy_in(files, **kwargs):
         dst = fspec.workdir or kwargs.get('workdir') or '.'
         logger.info('the file will be stored in %s' % str(dst))
 
+        # process the stage in
         error_msg = None
         rucio_state = None
         try:
-            rucio_state = _stage_in_api(dst, fspec, trace_report)
+            rucio_state = _stage_in_api(dst, fspec, trace_report, traces_rucio)
         except Exception as error:
             error_msg = str(error)
 
+        # treatment of error states
         if error_msg:
             logger.info('stderr = %s' % error_msg)
         else:
             logger.info('rucio client state: %s' % str(rucio_state))
 
-        if error_msg:  ## error occurred
+        if error_msg:
             error = resolve_common_transfer_errors(error_msg, is_stagein=True)
             fspec.status = 'failed'
             fspec.status_code = error.get('rcode')
             trace_report.update(clientState=rucio_state or 'STAGEIN_ATTEMPT_FAILED',
                                 stateReason=error_msg, timeEnd=time())
             if not ignore_errors:
+                if traces_rucio:
+                    _merge_traces([trace_report], traces_rucio)
                 trace_report.send()
                 raise PilotException(error_msg, code=error.get('rcode'), state='FAILED')
 
         # verify checksum; compare local checksum with catalog value (fspec.checksum), use same checksum type
+        # terminating with sending the trace
         destination = os.path.join(dst, fspec.lfn)
         if os.path.exists(destination):
             state, diagnostics = verify_catalog_checksum(fspec, destination)
             if diagnostics != "" and not ignore_errors:
                 trace_report.update(clientState=state or 'STAGEIN_ATTEMPT_FAILED', stateReason=diagnostics,
                                     timeEnd=time())
+                if traces_rucio:
```

Comment:
Tomas, 

why it's too much complicated and requires to merge traces?

in order merge 2 dictionary it will be enough  just to appy `dict.update`, we don't use nested data in the trace report,

why all this overhead implementation with `_merge_traces` and `trace.merge` is required?
when it can be done in one line.
```
extdata = {'x':'y'}
trace_report.update(extdata)
```
or 
```trace_report.update(extdata.copy())```

Reply:
Tomas can you check this? Now is a good time for making this update. Please also provide a one-liner that I can add to the pilot update message when we release it. Thanks